posts/2023-12-21-async-fn-rpit-in-traits.md

```diff
@@ -0,0 +1,229 @@
+
+## Frequently asked questions
+
+### Is it okay to use `-> impl Trait` in traits?
+
+For private traits you can use `-> impl Trait` freely. For public traits, it's best to avoid them for now unless you can anticipate all the bounds your users might want (in which case you can use `#[trait_variant]`, as we do for async). We expect to lift this restriction in the future.
+
+### Should I still use the `#[async_trait]` macro?
+
+There are a couple of reasons you might need to continue using async-trait:
+
+* You want to support Rust versions older than 1.75.
+* You want dynamic dispatch.
+
+As stated above, we hope to enable dynamic dispatch in a future version of trait-variant.
+
+### Is it okay to use `async fn` in traits? What are the limitations?
+
+Assuming you don't need to use `#[async_trait]` for one of the reasons stated above, it's totally fine to use regular `async fn` in traits. Just remember to use `#[trait_variant]` if you want to support multithreaded runtimes.
+
+The biggest limitation is that a type must always decide if it implements the Send or non-Send version of a trait. It cannot implement the Send version *conditionally* on one of its generics. This can come up in the [middleware](https://github.com/tower-rs/tower/blob/master/guides/building-a-middleware-from-scratch.md) pattern, for example, `RequestLimitingService<T>` that is HttpService if `T: HttpService`.
+
+### Why do I need `#[trait_variant]` and Send bounds?
+
+In simple cases you may find that your trait appears to work fine with a multithreaded executor. There are some patterns that just won't work, however. Consider the following:
+
+```rust!
+fn spawn_task(service: impl HttpService + 'static) {
+    tokio::spawn(async move {
+        let url = Url::from("https://rust-lang.org");
+        let _body = service.fetch(url).await;
+    });
+}
+```
+
+Without Send bounds on our trait, this would fail to compile with the error: "future cannot be sent between threads safely". By creating a variant of your trait with Send bounds, you avoid sending your users into this trap.
+
+### Can I mix async fn and impl trait?
+
+Yes, you can freely move between the `async fn` and `-> impl Future` spelling in your traits and impls. This is true even when one form has a Send bound.[^leakage] This makes the traits created by `trait_variant` nicer to use.
+
+```rust!
+trait HttpService: Send {
+    fn fetch(&self, url: Url)
+    -> impl Future<Output = HtmlBody> + Send;
+}
+
+impl HttpService for MyService {
+    async fn fetch(&self, url: Url) -> HtmlBody {
+        // This works, as long as `do_fetch(): Send`!
+        self.client.do_fetch(url).await.into_body()
+    }
+}
+```
+
+[^leakage]: This works because of *auto-trait leakage*, which allows knowledge of auto traits to "leak" outside of an impl signature that does not specify them.
```

Comment:
```suggestion
[^leakage]: This works because of *auto-trait leakage*, which allows knowledge of auto traits to "leak" from an item whose signature that does not specify them.
```